Optimize cornerMinEigenVal

komakai opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] The feature is well documented and sample code can be built with the project CMake

I noticed in the calcMinEigenValLine_AVX implementation that there is a lot of work goes into interleaving the x^2, xy and  y^2 values into a single Mat with 3 channels only to then deinterleave them again later.
By changing to use 3 Mats of 1 channel we can get rid of the interleaving/deinterleaving giving about a 5% speed up. Also then the AVX512 implementation becomes much easier.

Final thought on this - after my change the SIMD128, AVX256 and AVXX512 code all look nearly the same. @alalek - is there a way in OpenCV to write generic SIMD code and just let the compiler select the widest available vector operations for the target platform?


```
force_builders=Linux AVX2,Custom
buildworker:Custom=linux-4
build_image:Custom=ubuntu:18.04
CPU_BASELINE:Custom=AVX512_SKX
test_filter:Custom=none
disable_ipp=ON
```

alalek commented:
> widest available vector operations

By default these operations are available through `vx_` types and operations. However some of these operations are still not portable (need to rework masks support).
- https://docs.opencv.org/3.4/df/d91/group__core__hal__intrin.html
- tutorial is going to be added here: https://github.com/opencv/opencv/pull/20361/files#diff-fabe36c8737a5c3786c6783f5309addb211d651475c6df132e907ef69fec188bR67 ("Register Structures" section)

komakai commented:
> By default these operations are available through vx_ types and operations. However some of these operations are still not portable (need to rework masks support).

@alalek  thanks - will check these out

komakai commented:
@alalek I tried out using `vx_xxx` operations but even though I have AVX (256 bit vector support) they only give me 128 bit vectors.
<img width="360" alt="vx_load" src="https://user-images.githubusercontent.com/5875290/128617176-6aa3e3fb-5d21-4217-b313-0fea3d3d69f6.png">
Also need to consider things like the widest vector available at run time may be different from the widest vector at compile time. (Need to check what is available via `checkHardwareSupport`)

alalek commented:
Only AVX2 is supported through OpenCV SIMD intrinsics. Just AVX is not supported for now.

https://github.com/opencv/opencv/blob/4.5.3/modules/core/include/opencv2/core/hal/intrin.hpp#L249-L255

alalek commented on modules/imgproc/src/corner.avx.cpp:
```diff
@@ -49,132 +49,112 @@
-}
-
-int calcMinEigenValLine_AVX(const float* cov, float* dst, int width)
+int calcMinEigenValLine_AVX(const float* cov_x2, const float* cov_xy, const float* cov_y2, float* dst, int width)
```
Please note, than this file is named with `.avx.cpp` suffix. It should contain AVX optimization only. There is no place to support AVX512.

Generic support for different instruction sets are implemented through:
- `.simd.hpp`
- and `.dispatch.cpp`

files.

See also: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

komakai replied:
Thanks for the info @alalek 

> Please note, than this file is named with .avx.cpp suffix. It should contain AVX optimization only. There is no place to support AVX512.

I had actually worked this out and also found that by moving the AVX512 code to a file with extension `avx_512f.cpp` it could compile correctly with `-mavx512f` build flags etc. (Not sure if this intended because currently there are no source files in OpenCV with that file extension but it is possible - https://github.com/komakai/opencv/commit/f71fa77a4c22ac46dea836d240d3685460d747aa)

As per my previous comment - basically having 3 separate copies of the same algorithm doesn’t make much sense so I am continuing to try and find a way to make a single generic version like https://github.com/komakai/opencv/commit/cfbfea0966b0ff241d3772190edfbe9153908eb1

Unfortunately because only AVX2 is supported through OpenCV SIMD intrinsics currently, this actually results in worse performance on a machine with AVX but not AVX2. I guess the thing to do would be to add AVX support to OpenCV SIMD intrinsics - shouldn’t be too difficult - correct?

komakai commented:
Converting to draft while I think about how to proceed

asmorkalov commented:
@komakai Do you plan to finish the PR?

komakai commented:
> @komakai Do you plan to finish the PR?

Yes eventually - please note it is in Draft status. I will remove the Draft status when it is finished

komakai commented:
> @komakai Do you plan to finish the PR?

Sorry for the delay - it's ready now

@hanliutong - I made modification to my code to try to match your changes for universal intrinsics as below:
* changed to use **vx_setall_f32** instead of **vx_setall**
* changed to use **VTraits<v_float32>::vlanes()** instead of **v_float::nlanes**

Hopefully I haven't broken any of your changes but apologies in advance if I  have.

opencv-alalek commented:
> Optimize

Could you add a report (to PR description) from performance tests with and without this patch?

komakai commented:
> Could you add a report (to PR description) from performance tests with and without this patch?

I have run some tests but because the smallest unit of time that the tests can distinguish is milliseconds then its difficult to get an accurate value for the performance improvement. Is there any way to get test times in microseconds

komakai commented:
Turns out there is no difference in the code output by the AVX and AVX2 cases (I was thinking _mm256_fmadd_ps was an AVX2 instruction and there would be a small speed up because of that but it turns out it is actually in the FMA3 instruction set). I will update the PR to remove the AVX2 case.

opencv-alalek commented:
> Is there any way to get test times in microseconds?

Micro-benchmarks require more system stability requirements or provide unstable results.
Prefer to have results in ms, e.g. increase input size or run function 10 times (test name should be named accordingly)
